{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "## The Data Set\n",
    "In today's workshop, we will revisit the data set you worked with in the Machine Learning workshop. As a refresher:  this data set is from the GSE53987 dataset on Bipolar disorder (BD) and major depressive disorder (MDD) and schizophrenia:\n",
    "\n",
    "Lanz TA, Joshi JJ, Reinhart V, Johnson K et al. STEP levels are unchanged in pre-frontal cortex and associative striatum in post-mortem human brain samples from subjects with schizophrenia, bipolar disorder and major depressive disorder. PLoS One 2015;10(3):e0121744. PMID: 25786133\n",
    "\n",
    "This is a microarray data on platform GPL570 (HG-U133_Plus_2, Affymetrix Human Genome U133 Plus 2.0 Array) consisting of 54675 probes.\n",
    "\n",
    "The raw CEL files of the GEO series were downloaded, frozen-RMA normalized, and the probes have been converted to HUGO gene symbols using the annotate package averaging on genes. The sample clinical data (meta-data) was parsed from the series matrix file. You can download it **here**.  \n",
    "\n",
    "In total there are 205 rows consisting of 19 individuals diagnosed with BPD, 19 with MDD, 19 schizophrenia and 19 controls. Each sample has gene expression from 3 tissues (post-mortem brain). There are a total of 13768 genes (numeric features) and 10 meta features and 1 ID (GEO sample accession):\n",
    "\n",
    "- Age\n",
    "- Race (W for white and B for black)\n",
    "- Gender (F for female and M for male)\n",
    "- Ph: pH of the brain tissue\n",
    "- Pmi: post mortal interval\n",
    "- Rin: RNA integrity number\n",
    "- Patient: Unique ID for each patient. Each patient has up to 3 tissue samples. The patient ID is written as disease followed by a number from 1 to 19\n",
    "- Tissue: tissue the expression was obtained from.\n",
    "- Disease.state: class of disease the patient belongs to: bipolar, schizophrenia, depression or control.\n",
    "- source.name: combination of the tissue and disease.state\n",
    "\n",
    "## Workshop Goals\n",
    "This workshop will walk you through an analysis of the GSE53987 microarray data set. This workshop has the following three tasks:  \n",
    "    1. Visualize the demographics of the data set  \n",
    "    2. Cluster gene expression data and appropriately visualize the cluster results\n",
    "    3. Compute differential gene expression and visualize the differential expression\n",
    "\n",
    "Each task has a __required__ section and a __bonus__ section. Focus on completing the three __required__ sections first, then if you have time at the end, revisit the __bonus__ sections. \n",
    "\n",
    "Finally, as this is your final workshop, we hope that you will this as an opportunity to integrate the different concepts that you have learned in previous workshops.\n",
    "\n",
    "## Workshop Logistics\n",
    "As mentioned in the pre-workshop documentation, you can do this workshop either in a Jupyter Notebook, or in a python script. Please make sure you have set-up the appropriate environment for youself. This workshop will be completed using \"paired-programming\" and the \"driver\" will switch every 15 minutes. Also, we will be using the python plotting libraries matplotlib and seaborn. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TASK 0: Import Libraries and Data\n",
    "- Download the data set (above) as a .csv file\n",
    "- Initialize your script by loading the following libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Necessary Libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from sklearn import cluster, metrics, decomposition\n",
    "from matplotlib import pyplot as plt\n",
    "import itertools\n",
    "data = pd.read_csv('GSE53987_combined.csv', index_col=0)\n",
    "genes = data.columns[10:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TASK 1: Visualize Dataset Demographics  \n",
    "### Required Workshop Task:  \n",
    "##### Use the skeleton code to write 3 plotting functions:  \n",
    "    1. plot_distribution()  \n",
    "        - Returns a distribution plot object given a dataframe and one observation\n",
    "    2. plot_relational()\n",
    "        - Returns a distribution plot object given a dataframe and (x,y) observations  \n",
    "    3. plot_categorical()\n",
    "        - Returns a categorical plot object given a dataframe and (x,y) observations\n",
    "##### Use these functions to produce the following plots:\n",
    "    1. Histogram of patient ages\n",
    "    2. Histogram of gene expression for 1 gene\n",
    "    3. Scatter plot of gene expression for 1 gene by ages \n",
    "    4. Scatter plot of gene expression for 1 gene by disease state \n",
    "Your plots should satisfy the following critical components:  \n",
    "    - Axis titles\n",
    "    - Figure title\n",
    "    - Legend (if applicable)\n",
    "    - Be readable\n",
    "    \n",
    "### Bonus Task: \n",
    "1. Return to these functions and include functionality to customize color palettes, axis legends, etc. You can choose to define your own plotting \"style\" and keep that consistent for all of your plotting functions.  \n",
    "2. Faceting your plots. Modify your functions to take in a \"facet\" argument that when facet is an observation, the function will create a facet grid and facet on that observation. Read more about faceting here:\n",
    "Faceting generates multi-plot grids by __mapping a dataset onto multiple axes arrayed in a grid of rows and columns that correspond to levels of variables in the dataset.__  \n",
    "    - In order to use facteting, your data __must be__ in a Pandas DataFrame and it must take the form of what Hadley Whickam calls “tidy” data. \n",
    "    - In brief, that means your dataframe should be structured such that each column is a variable and each row is an observation. There are figure-level functions (e.g. relplot() or catplot()) that will create facet grids automatically and can be used in place of things like distplot() or scatterplot(). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the data (.csv file) as a data frame\n",
    "data = pd.read_csv(\"/Users/ebriars/Desktop/Bioinformatics/BRITE REU Workshops/Data Visualization/GSE53987_combined.csv\", index_col=0)\n",
    "\n",
    "# Function to Plot a Distribtion\n",
    "def plot_distribution(df, obs1, obs2=''):\n",
    "    \"\"\"\n",
    "    Create a distribution plot for at least one observation\n",
    "    \n",
    "    Arguments:\n",
    "        df (pandas data frame): data frame containing at least 1 column of numerical values\n",
    "        obs1 (string): observation to plot distribution on\n",
    "        obs2 (string, optional)\n",
    "    Returns:\n",
    "        axes object\n",
    "    \"\"\"\n",
    "    if obs2 == '':\n",
    "        ax = sns.distplot(df[obs1])\n",
    "    else:\n",
    "        ax = sns.FacetGrid(df, hue=obs2)\n",
    "        ax = (g.map(sns.distplot, obs1, hist=False))\n",
    "    return ax\n",
    "\n",
    "# Function to Plot Relational (x,y) Plots \n",
    "def plot_relational(df, x, y, hue=None, kind=None):\n",
    "    \"\"\"\n",
    "    Create a plot for an x,y relationship (default = scatter plot)\n",
    "    Optional functionality for additional observations.\n",
    "    \n",
    "    Arguments:\n",
    "        df (pandas data frame): data frame containing at least 2 columns of numerical values\n",
    "        x (string): observation for the independent variable\n",
    "        y (string): observation for the dependent variable\n",
    "        hue (string, optional): additional observation to color the plot on\n",
    "        kind (string, optional): type of plot to create [scatter, line]\n",
    "    Returns:\n",
    "        axes object\n",
    "    \"\"\"\n",
    "    if kind == None or kind == \"scatter\":\n",
    "        ax = sns.scatterplot(data=df, x=x, y=y, hue=hue)\n",
    "    else:\n",
    "        ax = sns.lineplot(data=df, x=x, y=y, hue=hue)\n",
    "    return ax\n",
    "\n",
    "def plot_categorical(df, x, y, hue=None, kind=None):\n",
    "    \"\"\"\n",
    "    Create a plot for an x,y relationship where x is categorical (not numerical)\n",
    "    \n",
    "    Arguments:\n",
    "        df (pandas data frame): data frame containing at least 2 columns of numerical values\n",
    "        x (string): observation for the independent variable (categorical)\n",
    "        y (string): observation for the dependent variable\n",
    "        hue (string, optional): additional observation to color the plot on\n",
    "        kind (string, optional): type of plot to create. Options should include at least: \n",
    "        strip (default), box, and violin\n",
    "    \"\"\"\n",
    "    if kind == None or kind == \"strip\":\n",
    "        ax = sns.stripplot(data=df, x=x, y=y, hue=hue)\n",
    "    elif kind == \"violin\":\n",
    "        ax = sns.violinplot(data=df, x=x, y=y, hue=hue)\n",
    "    elif kind == \"box\":\n",
    "        ax = sns.boxplot(data=df, x=x, y=y, hue=hue)\n",
    "    return ax\n",
    "\n",
    "def main():\n",
    "    \"\"\"\n",
    "    Generate the following plots:\n",
    "    1. Histogram of patient ages\n",
    "    2. Histogram of gene expression for 1 gene\n",
    "    3. Scatter plot of gene expression for 1 gene by ages \n",
    "    4. Scatter plot of gene expression for 1 gene by disease state \n",
    "    \"\"\"\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TASK 2: Differential Expression Analysis\n",
    "\n",
    "Differential expression analysis is a fancy way of saying, \"We want to find which genes exhibit increased or decreased expression compared to a control group\". Neat. Because the dataset we're working with is MicroArray data -- which is mostly normally distributed -- we'll be using a simple One-Way ANOVA. If, however, you were working with sequence data -- which follows a Negative Binomial distribution -- you would need more specialized tools. A helper function is provided below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def differential_expression(data, group_col, features, reference=None):\n",
    "    \"\"\"\n",
    "    Perform a one-way ANOVA across all provided features for a given grouping.\n",
    "    \n",
    "    Arguments\n",
    "    ---------\n",
    "    \n",
    "        data : (pandas.DataFrame)\n",
    "            DataFrame containing group information and feature values.\n",
    "        group_col : (str)\n",
    "            Column in `data` containing sample group labels.\n",
    "        features : (list, numpy.ndarray):\n",
    "            Columns in `data` to test for differential expression. Having them\n",
    "            be gene names would make sense. :thinking:\n",
    "        reference : (str, optional)\n",
    "            Value in `group_col` to use as the reference group. Default is None,\n",
    "            and the value will be chosen.\n",
    "            \n",
    "    Returns\n",
    "    -------\n",
    "        pandas.DataFrame\n",
    "            A DataFrame of differential expression results with columns for\n",
    "            fold changes between groups, maximum fold change from reference,\n",
    "            f values, p values, and adjusted p-values by Bonferroni correction.\n",
    "    \"\"\"\n",
    "    if group_col not in data.columns:\n",
    "        raise ValueError(\"`group_col` {} not found in data\".format(group_col))\n",
    "    if any([x not in data.columns for x in features]):\n",
    "        raise ValueError(\"Not all provided features found in data.\")\n",
    "    if reference is None:\n",
    "        reference = data[group_col].unique()[0]\n",
    "        print(\"No reference group provided. Using {}\".format(reference))\n",
    "    elif reference not in data[group_col].unique():\n",
    "        raise ValueError(\"Reference value {} not found in column {}.\".format(\n",
    "                         reference, group_col))\n",
    "    by_group = data.groupby(group_col)\n",
    "    reference_avg = by_group.get_group(reference).loc[:,features].mean()\n",
    "    values = []\n",
    "    results = {}\n",
    "    for each, index in by_group.groups.items():\n",
    "        values.append(data.loc[index, features])\n",
    "        if each !=  reference:\n",
    "            key = \"{}.FoldChange\".format(each)\n",
    "            results[key] = data.loc[index, features].mean()\\\n",
    "                         / reference_avg\n",
    "    fold_change_cols = list(results.keys())\n",
    "    fvalues, pvalues = stats.f_oneway(*values)\n",
    "    results['f.value'] = fvalues\n",
    "    results['p.value'] = pvalues\n",
    "    results['p.value.adj'] = pvalues * len(features)\n",
    "    results_df = pd.DataFrame(results)\n",
    "    def largest_deviation(x):\n",
    "        i = np.where(abs(x) == max(abs(x)))[0][0]\n",
    "        return x[i]\n",
    "    results_df['Max.FoldChange'] = results_df[fold_change_cols].apply(\n",
    "                                       lambda x: largest_deviation(x.values), axis=1)\n",
    "\n",
    "    return results_df  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Source.name</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Race</th>\n",
       "      <th>Pmi</th>\n",
       "      <th>Ph</th>\n",
       "      <th>Rin</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Disease.state</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM8.AS1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1304870</th>\n",
       "      <td>control_10</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>51</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>24.20</td>\n",
       "      <td>6.6</td>\n",
       "      <td>7.8</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.901962</td>\n",
       "      <td>6.924233</td>\n",
       "      <td>4.883176</td>\n",
       "      <td>7.037155</td>\n",
       "      <td>4.221055</td>\n",
       "      <td>4.071727</td>\n",
       "      <td>6.101329</td>\n",
       "      <td>6.840671</td>\n",
       "      <td>5.802425</td>\n",
       "      <td>7.951339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304871</th>\n",
       "      <td>control_11</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>51</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>7.80</td>\n",
       "      <td>6.6</td>\n",
       "      <td>7.2</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.891827</td>\n",
       "      <td>6.686104</td>\n",
       "      <td>4.919824</td>\n",
       "      <td>7.019892</td>\n",
       "      <td>4.449879</td>\n",
       "      <td>4.110077</td>\n",
       "      <td>6.372282</td>\n",
       "      <td>7.096796</td>\n",
       "      <td>5.972458</td>\n",
       "      <td>7.995525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304872</th>\n",
       "      <td>control_12</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>36</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>14.50</td>\n",
       "      <td>6.4</td>\n",
       "      <td>8.0</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.162939</td>\n",
       "      <td>7.245197</td>\n",
       "      <td>4.950240</td>\n",
       "      <td>7.558985</td>\n",
       "      <td>4.500566</td>\n",
       "      <td>4.214102</td>\n",
       "      <td>6.515646</td>\n",
       "      <td>6.374378</td>\n",
       "      <td>5.741482</td>\n",
       "      <td>7.969294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304873</th>\n",
       "      <td>control_13</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>18.50</td>\n",
       "      <td>6.5</td>\n",
       "      <td>7.0</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.926336</td>\n",
       "      <td>6.451683</td>\n",
       "      <td>4.352600</td>\n",
       "      <td>6.130494</td>\n",
       "      <td>4.395073</td>\n",
       "      <td>4.175186</td>\n",
       "      <td>6.513069</td>\n",
       "      <td>7.194028</td>\n",
       "      <td>5.851711</td>\n",
       "      <td>7.977704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304874</th>\n",
       "      <td>control_14</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>55</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>28.00</td>\n",
       "      <td>6.1</td>\n",
       "      <td>6.8</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.973318</td>\n",
       "      <td>6.760423</td>\n",
       "      <td>5.264255</td>\n",
       "      <td>7.481936</td>\n",
       "      <td>4.387548</td>\n",
       "      <td>4.218758</td>\n",
       "      <td>6.709437</td>\n",
       "      <td>6.740813</td>\n",
       "      <td>5.929435</td>\n",
       "      <td>7.820045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304875</th>\n",
       "      <td>control_15</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>22</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>20.10</td>\n",
       "      <td>6.8</td>\n",
       "      <td>7.1</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.860300</td>\n",
       "      <td>6.748603</td>\n",
       "      <td>4.718314</td>\n",
       "      <td>7.282907</td>\n",
       "      <td>4.234804</td>\n",
       "      <td>4.009726</td>\n",
       "      <td>6.335518</td>\n",
       "      <td>6.874673</td>\n",
       "      <td>6.012384</td>\n",
       "      <td>7.792208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304876</th>\n",
       "      <td>control_16</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>52</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>22.60</td>\n",
       "      <td>7.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.233145</td>\n",
       "      <td>6.715545</td>\n",
       "      <td>4.307412</td>\n",
       "      <td>6.938351</td>\n",
       "      <td>4.361942</td>\n",
       "      <td>4.094315</td>\n",
       "      <td>6.422520</td>\n",
       "      <td>6.849899</td>\n",
       "      <td>5.829819</td>\n",
       "      <td>8.066799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304877</th>\n",
       "      <td>control_17</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>58</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>22.70</td>\n",
       "      <td>6.4</td>\n",
       "      <td>6.3</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.029479</td>\n",
       "      <td>6.418220</td>\n",
       "      <td>4.516272</td>\n",
       "      <td>6.932229</td>\n",
       "      <td>4.317337</td>\n",
       "      <td>4.087772</td>\n",
       "      <td>6.366974</td>\n",
       "      <td>7.096214</td>\n",
       "      <td>6.021788</td>\n",
       "      <td>7.811885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304878</th>\n",
       "      <td>control_18</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>40</td>\n",
       "      <td>F</td>\n",
       "      <td>B</td>\n",
       "      <td>16.60</td>\n",
       "      <td>6.8</td>\n",
       "      <td>7.9</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.925501</td>\n",
       "      <td>6.953431</td>\n",
       "      <td>4.801928</td>\n",
       "      <td>7.410219</td>\n",
       "      <td>4.367789</td>\n",
       "      <td>4.151118</td>\n",
       "      <td>6.216399</td>\n",
       "      <td>6.737730</td>\n",
       "      <td>5.838513</td>\n",
       "      <td>7.889143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304879</th>\n",
       "      <td>control_19</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>41</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>15.40</td>\n",
       "      <td>6.6</td>\n",
       "      <td>8.5</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.946981</td>\n",
       "      <td>6.942262</td>\n",
       "      <td>5.144930</td>\n",
       "      <td>7.331290</td>\n",
       "      <td>4.142431</td>\n",
       "      <td>4.094786</td>\n",
       "      <td>6.351937</td>\n",
       "      <td>6.892228</td>\n",
       "      <td>5.911674</td>\n",
       "      <td>7.778130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304880</th>\n",
       "      <td>control_1</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>49</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>21.20</td>\n",
       "      <td>6.5</td>\n",
       "      <td>7.8</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.992757</td>\n",
       "      <td>6.677630</td>\n",
       "      <td>4.621341</td>\n",
       "      <td>7.638745</td>\n",
       "      <td>4.379308</td>\n",
       "      <td>4.075940</td>\n",
       "      <td>6.323545</td>\n",
       "      <td>6.912604</td>\n",
       "      <td>6.044937</td>\n",
       "      <td>7.999910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304881</th>\n",
       "      <td>control_2</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>48</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>21.68</td>\n",
       "      <td>6.6</td>\n",
       "      <td>7.3</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.519082</td>\n",
       "      <td>6.926018</td>\n",
       "      <td>5.272251</td>\n",
       "      <td>7.182019</td>\n",
       "      <td>4.105560</td>\n",
       "      <td>4.069822</td>\n",
       "      <td>6.569197</td>\n",
       "      <td>6.609872</td>\n",
       "      <td>6.054169</td>\n",
       "      <td>8.018231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304882</th>\n",
       "      <td>control_3</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>39</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>24.50</td>\n",
       "      <td>6.8</td>\n",
       "      <td>8.2</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.935350</td>\n",
       "      <td>6.586736</td>\n",
       "      <td>5.029396</td>\n",
       "      <td>6.951266</td>\n",
       "      <td>4.281081</td>\n",
       "      <td>4.063719</td>\n",
       "      <td>6.366720</td>\n",
       "      <td>6.785588</td>\n",
       "      <td>5.893738</td>\n",
       "      <td>7.831109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304883</th>\n",
       "      <td>control_4</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>48</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>24.50</td>\n",
       "      <td>6.5</td>\n",
       "      <td>7.0</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.767633</td>\n",
       "      <td>6.590179</td>\n",
       "      <td>4.663711</td>\n",
       "      <td>7.618855</td>\n",
       "      <td>4.450818</td>\n",
       "      <td>4.210372</td>\n",
       "      <td>6.185696</td>\n",
       "      <td>7.042597</td>\n",
       "      <td>5.844226</td>\n",
       "      <td>7.814933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304884</th>\n",
       "      <td>control_5</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>43</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>13.80</td>\n",
       "      <td>6.6</td>\n",
       "      <td>7.6</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.094495</td>\n",
       "      <td>6.722968</td>\n",
       "      <td>5.064823</td>\n",
       "      <td>7.388035</td>\n",
       "      <td>4.231063</td>\n",
       "      <td>4.027637</td>\n",
       "      <td>6.586934</td>\n",
       "      <td>6.543934</td>\n",
       "      <td>5.800227</td>\n",
       "      <td>8.067118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304885</th>\n",
       "      <td>control_6</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>68</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>11.80</td>\n",
       "      <td>6.8</td>\n",
       "      <td>6.1</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.459846</td>\n",
       "      <td>6.608134</td>\n",
       "      <td>3.801517</td>\n",
       "      <td>6.415573</td>\n",
       "      <td>4.325173</td>\n",
       "      <td>4.072440</td>\n",
       "      <td>6.465788</td>\n",
       "      <td>7.302735</td>\n",
       "      <td>5.867902</td>\n",
       "      <td>7.841796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304886</th>\n",
       "      <td>control_7</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>58</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>18.80</td>\n",
       "      <td>6.6</td>\n",
       "      <td>7.2</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>6.282578</td>\n",
       "      <td>6.809128</td>\n",
       "      <td>4.388218</td>\n",
       "      <td>7.243010</td>\n",
       "      <td>4.336328</td>\n",
       "      <td>4.165654</td>\n",
       "      <td>6.419830</td>\n",
       "      <td>7.030138</td>\n",
       "      <td>6.069576</td>\n",
       "      <td>8.086008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1304887</th>\n",
       "      <td>control_8</td>\n",
       "      <td>hippocampus, control</td>\n",
       "      <td>43</td>\n",
       "      <td>M</td>\n",
       "      <td>W</td>\n",
       "      <td>22.30</td>\n",
       "      <td>6.7</td>\n",
       "      <td>7.9</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>control</td>\n",
       "      <td>...</td>\n",
       "      <td>5.998734</td>\n",
       "      <td>6.615258</td>\n",
       "      <td>5.077771</td>\n",
       "      <td>7.319819</td>\n",
       "      <td>4.393906</td>\n",
       "      <td>4.183219</td>\n",
       "      <td>6.793573</td>\n",
       "      <td>6.736603</td>\n",
       "      <td>5.948742</td>\n",
       "      <td>8.153599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18 rows × 13778 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Patient           Source.name  Age Gender Race    Pmi   Ph  \\\n",
       "GSM1304870  control_10  hippocampus, control   51      M    W  24.20  6.6   \n",
       "GSM1304871  control_11  hippocampus, control   51      F    W   7.80  6.6   \n",
       "GSM1304872  control_12  hippocampus, control   36      F    W  14.50  6.4   \n",
       "GSM1304873  control_13  hippocampus, control   65      F    W  18.50  6.5   \n",
       "GSM1304874  control_14  hippocampus, control   55      M    W  28.00  6.1   \n",
       "GSM1304875  control_15  hippocampus, control   22      M    W  20.10  6.8   \n",
       "GSM1304876  control_16  hippocampus, control   52      F    W  22.60  7.1   \n",
       "GSM1304877  control_17  hippocampus, control   58      F    W  22.70  6.4   \n",
       "GSM1304878  control_18  hippocampus, control   40      F    B  16.60  6.8   \n",
       "GSM1304879  control_19  hippocampus, control   41      F    W  15.40  6.6   \n",
       "GSM1304880   control_1  hippocampus, control   49      M    W  21.20  6.5   \n",
       "GSM1304881   control_2  hippocampus, control   48      M    W  21.68  6.6   \n",
       "GSM1304882   control_3  hippocampus, control   39      F    W  24.50  6.8   \n",
       "GSM1304883   control_4  hippocampus, control   48      M    W  24.50  6.5   \n",
       "GSM1304884   control_5  hippocampus, control   43      M    W  13.80  6.6   \n",
       "GSM1304885   control_6  hippocampus, control   68      M    W  11.80  6.8   \n",
       "GSM1304886   control_7  hippocampus, control   58      F    W  18.80  6.6   \n",
       "GSM1304887   control_8  hippocampus, control   43      M    W  22.30  6.7   \n",
       "\n",
       "            Rin       Tissue Disease.state  ...  ZSWIM8.AS1      ZW10  \\\n",
       "GSM1304870  7.8  hippocampus       control  ...    5.901962  6.924233   \n",
       "GSM1304871  7.2  hippocampus       control  ...    5.891827  6.686104   \n",
       "GSM1304872  8.0  hippocampus       control  ...    6.162939  7.245197   \n",
       "GSM1304873  7.0  hippocampus       control  ...    5.926336  6.451683   \n",
       "GSM1304874  6.8  hippocampus       control  ...    5.973318  6.760423   \n",
       "GSM1304875  7.1  hippocampus       control  ...    5.860300  6.748603   \n",
       "GSM1304876  7.0  hippocampus       control  ...    6.233145  6.715545   \n",
       "GSM1304877  6.3  hippocampus       control  ...    6.029479  6.418220   \n",
       "GSM1304878  7.9  hippocampus       control  ...    5.925501  6.953431   \n",
       "GSM1304879  8.5  hippocampus       control  ...    5.946981  6.942262   \n",
       "GSM1304880  7.8  hippocampus       control  ...    5.992757  6.677630   \n",
       "GSM1304881  7.3  hippocampus       control  ...    5.519082  6.926018   \n",
       "GSM1304882  8.2  hippocampus       control  ...    5.935350  6.586736   \n",
       "GSM1304883  7.0  hippocampus       control  ...    5.767633  6.590179   \n",
       "GSM1304884  7.6  hippocampus       control  ...    6.094495  6.722968   \n",
       "GSM1304885  6.1  hippocampus       control  ...    6.459846  6.608134   \n",
       "GSM1304886  7.2  hippocampus       control  ...    6.282578  6.809128   \n",
       "GSM1304887  7.9  hippocampus       control  ...    5.998734  6.615258   \n",
       "\n",
       "              ZWILCH     ZWINT      ZXDA      ZXDB      ZXDC       ZYX  \\\n",
       "GSM1304870  4.883176  7.037155  4.221055  4.071727  6.101329  6.840671   \n",
       "GSM1304871  4.919824  7.019892  4.449879  4.110077  6.372282  7.096796   \n",
       "GSM1304872  4.950240  7.558985  4.500566  4.214102  6.515646  6.374378   \n",
       "GSM1304873  4.352600  6.130494  4.395073  4.175186  6.513069  7.194028   \n",
       "GSM1304874  5.264255  7.481936  4.387548  4.218758  6.709437  6.740813   \n",
       "GSM1304875  4.718314  7.282907  4.234804  4.009726  6.335518  6.874673   \n",
       "GSM1304876  4.307412  6.938351  4.361942  4.094315  6.422520  6.849899   \n",
       "GSM1304877  4.516272  6.932229  4.317337  4.087772  6.366974  7.096214   \n",
       "GSM1304878  4.801928  7.410219  4.367789  4.151118  6.216399  6.737730   \n",
       "GSM1304879  5.144930  7.331290  4.142431  4.094786  6.351937  6.892228   \n",
       "GSM1304880  4.621341  7.638745  4.379308  4.075940  6.323545  6.912604   \n",
       "GSM1304881  5.272251  7.182019  4.105560  4.069822  6.569197  6.609872   \n",
       "GSM1304882  5.029396  6.951266  4.281081  4.063719  6.366720  6.785588   \n",
       "GSM1304883  4.663711  7.618855  4.450818  4.210372  6.185696  7.042597   \n",
       "GSM1304884  5.064823  7.388035  4.231063  4.027637  6.586934  6.543934   \n",
       "GSM1304885  3.801517  6.415573  4.325173  4.072440  6.465788  7.302735   \n",
       "GSM1304886  4.388218  7.243010  4.336328  4.165654  6.419830  7.030138   \n",
       "GSM1304887  5.077771  7.319819  4.393906  4.183219  6.793573  6.736603   \n",
       "\n",
       "               ZZEF1      ZZZ3  \n",
       "GSM1304870  5.802425  7.951339  \n",
       "GSM1304871  5.972458  7.995525  \n",
       "GSM1304872  5.741482  7.969294  \n",
       "GSM1304873  5.851711  7.977704  \n",
       "GSM1304874  5.929435  7.820045  \n",
       "GSM1304875  6.012384  7.792208  \n",
       "GSM1304876  5.829819  8.066799  \n",
       "GSM1304877  6.021788  7.811885  \n",
       "GSM1304878  5.838513  7.889143  \n",
       "GSM1304879  5.911674  7.778130  \n",
       "GSM1304880  6.044937  7.999910  \n",
       "GSM1304881  6.054169  8.018231  \n",
       "GSM1304882  5.893738  7.831109  \n",
       "GSM1304883  5.844226  7.814933  \n",
       "GSM1304884  5.800227  8.067118  \n",
       "GSM1304885  5.867902  7.841796  \n",
       "GSM1304886  6.069576  8.086008  \n",
       "GSM1304887  5.948742  8.153599  \n",
       "\n",
       "[18 rows x 13778 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's some pre-subsetted data\n",
    "hippocampus = data[data[\"Tissue\"] == \"hippocampus\"]\n",
    "pf_cortex = data[data[\"Tissue\"] == \"Pre-frontal cortex (BA46)\"]\n",
    "as_striatum = data[data[\"Tissue\"] == \"Associative striatum\"]\n",
    "# Here's how we can subset a dataset by two conditions.\n",
    "# You might find it useful :thinking:\n",
    "data[(data[\"Tissue\"] == 'hippocampus') & (data['Disease.state'] == 'control')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2a: Volcano Plots\n",
    "\n",
    "Volcano plots are ways to showcase the number of differentially expressed genes found during high throughput sequencing analysis. Log fold changes are plotted along the x-axis, while p-values are plotted along the y-axis. Genes are marked significant if they exceed some absolute Log fold change theshold **as well** some p-value level for significance. This can be seen in the plot below.\n",
    "\n",
    "![](https://galaxyproject.github.io/training-material/topics/transcriptomics/images/rna-seq-viz-with-volcanoplot/volcanoplot.png)\n",
    "\n",
    "Your first task will be to generate some Volcano plots:\n",
    "\n",
    "**Requirments**\n",
    "1. Use the provided function to perform an ANOVA (analysis of variance) between control and experimental groups in each tissue.\n",
    "    - Perform a separate analysis for each tissue.\n",
    "2. Implement the skeleton function to create a volcano plot to visualize both the log fold change in expression values and the adjusted p-values from the ANOVA\n",
    "3. Highlight significant genes with distinct colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def volcano_plot(data, sig_col, fc_col, sig_thresh, fc_thresh):\n",
    "    \"\"\"\n",
    "    Generate a volcano plot to showcasing differentially expressed genes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "        data : (pandas.DataFrame)\n",
    "            A data frame containing differential expression results\n",
    "        sig_col : str\n",
    "            Column in `data` with adjusted p-values.\n",
    "        fc_col : str\n",
    "            Column in `data` with fold changes.\n",
    "        sig_thresh : str\n",
    "            Threshold for statistical significance.\n",
    "        fc_thresh\n",
    "    \"\"\"\n",
    "    data['significant'] = False\n",
    "    data[fc_col] = np.log2(data[fc_col])\n",
    "    de_genes = (data[sig_col] < sig_thesh) & (data[fc_col].abs() > fc_thresh)\n",
    "    data.loc[de_genes, 'significant'] = True\n",
    "    ax = sns.scatterplot(x=fc_col, y=sig_col, hue='significant', data=data,\n",
    "                         palette=['black', 'red'], alpha=0.75)\n",
    "    linewidth = plt.rcParams['lines.linewidth'] - 1\n",
    "    plt.axvline(x=fc_thresh, linestyle='--', linewidth=linewidth,\n",
    "                color='#4D4E4F')\n",
    "    plt.axvline(x=-fc_thresh, linestyle='--', linewidth=linewidth,\n",
    "                color='#4D4E4F')\n",
    "    plt.axhline(y=sig_thresh, linestyle='--', linewidth=linewidth,\n",
    "                color='#4D4E4F')\n",
    "    ax.legend().set_visible(False)\n",
    "    ylabel = sig_col\n",
    "    if sig_col.lower() == 'fdr':\n",
    "        ylabel = 'False Discovery Rate'\n",
    "    plt.xlabel(r\"$log_2$ Fold Change\")\n",
    "    plt.ylabel(ylabel)\n",
    "    for spine in ['right', 'top']:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    plt.tight_layout()\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2b: Plot the Top 1000 Differentially Expressed Genes\n",
    "\n",
    "Clustered heatmaps are hugely popular for displaying differences in gene expression values. To reference such a plot, look back at the introductory material. Here we will be plotting the 1000 most differentially expressed genes for each of the analysis performed before.\n",
    "\n",
    "**Requirements**\n",
    "- Implement the skeleton function below\n",
    "- Z normalize gene values\n",
    "- Use a diverging and perceptually uniform colormap\n",
    "- Generate plots for each of the DE results above\n",
    "\n",
    "**Hint**: Look over all the options for [sns.clustermap()](https://seaborn.pydata.org/generated/seaborn.clustermap.html). It might make things easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(data, genes, group_col):\n",
    "    \"\"\"[summary]\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    data : pd.DataFrame\n",
    "        A (sample x gene) data matrix containing gene expression values for each sample.\n",
    "    genes : list, str\n",
    "        List of genes to plot   \n",
    "    \"\"\"\n",
    "\n",
    "    plot_data = anno_df[:, genes]\n",
    "    ax = sns.clustermap(data, cmap='RdBu_r', z_score=1)\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Bonus** There's nothing denoting which samples belong to which experimental group. Fix it.\n",
    "\n",
    "*Bonus hint*: Look real close at the documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TASK 3: Clustering Analysis\n",
    "\n",
    "You've seen clustering in the previous machine learning workshop. Some basic plots were generated for you, including plotting the clusters on the principle componets. While we can certainly do more of that, we will also be introducing two new plots: elbow plots and silhouette plots.\n",
    "\n",
    "### Elbow Plots\n",
    "\n",
    "Elbow plots are plots that are used to help diagnose the perennial question of K-means clustering: how do I chose K? To create the graph, you plot the number of clusters on the x-axis and some evaluation of \"cluster goodness\" on the y-axis. Looking at the name of the plot, you might guess that we're looking for an \"elbow\". This is the point in the graph when we start getting diminished returns in performance, and specifying more clusters may lead to over-clustering the data. An example plot is shown below.\n",
    "\n",
    "![](https://upload.wikimedia.org/wikipedia/commons/c/cd/DataClustering_ElbowCriterion.JPG)\n",
    "\n",
    "You can see the K selected (K = 3), is right before diminishing returns start to kick in. Mathematically, this point is defined as the point in which curvature is maximized. However, the inflection point is also a decent -- though more conservative -- estimate. However, we'll just stick to eye-balling it for this workshop. If you would like to know how to automatically find the elbow point, more information can be found [here](https://raghavan.usc.edu/papers/kneedle-simplex11.pdf)\n",
    "\n",
    "### Task 2a: Implement a function that creates an elbow plot\n",
    "\n",
    "Skeleton code is provided below. The function expects a list of k-values and their associated scores. An optional \"ax\" parameter is also provided. This parameter should be an axes object and can be created by issueing the following command:\n",
    "\n",
    "```ax = plt.subplot()```\n",
    "\n",
    "While we won't need the parameter right now, we'll likely use it in the future.\n",
    "\n",
    "**Function Requirements**\n",
    "- Generate plot data by clustering the entire dataset on the first 50 principle components. Vary K values from 2 - 10.\n",
    "    - While you've been supplied a helper function for clustering, you'll need to supply the principle components yourself. Refer to your machine learning workshop along with the scikit-learn [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)\n",
    "- Plots each k and it's associated value.\n",
    "- Plots lines connecting each data point.\n",
    "- Produces a plot with correctly labelled axes.\n",
    "\n",
    "**Hint:** Working with an axis object is similar to base matplotlib, except `plt.scatter()` might become something like `ax.scatter()`.\n",
    "\n",
    "#### Helper Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_data(X, k):\n",
    "    \"\"\"\n",
    "    Cluster data using K-Means.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "        X : (numpy.ndarray)\n",
    "            Data matrix to cluster samples on. Should be (samples x features).\n",
    "        k : int\n",
    "            Number of clusters to find.\n",
    "    Returns\n",
    "    -------\n",
    "        tuple (numpy.ndarray, float)\n",
    "            A tuple where the first value is the assigned cluster labels for\n",
    "            each sample, and the second value is the score associated with\n",
    "            the particular clustering.\n",
    "    \"\"\"\n",
    "    model = cluster.KMeans(n_clusters=k).fit(X)\n",
    "    score = model.score(X)\n",
    "    return (model.labels_, score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Task 2a Implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def elbow_plot(ks, scores, best=None, ax=None):\n",
    "    \"\"\"\n",
    "    Create a scatter plot to aid in choosing the number of clusters using K-means.\n",
    "    \n",
    "    \n",
    "    Arguments\n",
    "    ---------\n",
    "        ks : (numpy.ndarray)\n",
    "            Tested values for the number of clusters.\n",
    "        scores: (numpy.ndarray)\n",
    "            Cluster scores associated with each number K.\n",
    "        ax: plt.Axes Object, optional\n",
    "    \"\"\"\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots()\n",
    "    ax.scatter(ks, scores)\n",
    "    ax.plot(ks, scores)\n",
    "    ax.set_xlabel(\"Number of Clusters\")\n",
    "    ax.set_ylabel(\"Negative Distance From Mean\")\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once you've created the base plotting function, you'll probably realize we have no indivation of where the elbow point is. Fix this by adding another optional parameter (`best`) to your function. The parameter `best` should be the K value that produces the elbow point.\n",
    "\n",
    "**Function Requirements**\n",
    "\n",
    "- Add an optional parameter `best` that if supplied denotes the elbow point with a vertical, dashed line.\n",
    "- If `best` is not supplied, the plot should still be produced but without denoting the elbow point.\n",
    "\n",
    "**Hint**: `plt.axvline` and `plt.axhline` can be used to produce vertical and horizontal lines, respectively. More information [here](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.axvline.html)\n",
    "\n",
    "**Note**: You are not required to have the line end at the associated score value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def elbow_plot(ks, scores, best=None, ax=None):\n",
    "    \"\"\"\n",
    "    Create a scatter plot to aid in choosing the number of clusters using K-means.\n",
    "    \n",
    "    \n",
    "    Arguments\n",
    "    ---------\n",
    "        ks : (numpy.ndarray)\n",
    "            Tested values for the number of clusters.\n",
    "        scores: (numpy.ndarray)\n",
    "            Cluster scores associated with each number K.\n",
    "        best: int, optional\n",
    "            The best value for K. Determined by the K that falls at the elbow. If\n",
    "            passed, a black dashed line will be plotted to indicate the best.\n",
    "            Default is no line.\n",
    "        ax: plt.Axes Object, optional\n",
    "    \"\"\"\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots()\n",
    "    ax.scatter(ks, scores)\n",
    "    ax.plot(ks, scores)\n",
    "    ax.set_xlabel(\"Number of Clusters\")\n",
    "    ax.set_ylabel(\"Negative Distance From Mean\")\n",
    "    if best is not None:\n",
    "        if best not in ks:\n",
    "            raise ValueError(\"{} not included in provided number \"\n",
    "                             \"of clusters.\")\n",
    "        idx = np.where(np.array(ks) == best)[0][0]\n",
    "        ymin, ymax = ax.get_ylim()\n",
    "        point = (scores[idx] - ymin) / (ymax - ymin)\n",
    "        print(point)\n",
    "        ax.axvline(x=best, ymax=point, linestyle=\"--\", c='black')\n",
    "        ax.scatter([3], scores[idx], edgecolor='black', facecolor=\"none\",\n",
    "                    s=200)\n",
    "        ax.set_title(\"Elbow at K={}\".format(best), loc='left')\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Silhouette Plots\n",
    "\n",
    "Silhouette plots are another way to visually diagnose cluster performance. They are created by finding the [silhouette coefficient](https://en.wikipedia.org/wiki/Silhouette_(clustering)) for each sample in the data, and plotting an area graph for each cluster. The silhouette coefficient measures how well-separated clusters are from each other. The value ranges from $[-1 , 1]$, where 1 indicates good separation, 0 indicates randomness, and -1 indicates mixing of clusters.  An example is posted below.\n",
    "\n",
    "![](https://scikit-plot.readthedocs.io/en/stable/_images/plot_silhouette.png)\n",
    "\n",
    "As you can see, each sample in each cluster has the area filled from some minimal point (usually 0 or the minimum score in the dataset) and clusters are separated to produce distinct [silhouettes](https://www.youtube.com/watch?v=-TcUvXzgwMY).\n",
    "\n",
    "### Task 3b: Implement a function to plot silhouette coefficients\n",
    "\n",
    "Because the code for create a silhouette plot can be a little bit involved, we've created both a skeleton function with documentation, and provided the following pseudo-code:\n",
    "\n",
    "```\n",
    "- Calculate scores for each sample.\n",
    "- Get a set of unique sample labels.\n",
    "- Set a score minimum\n",
    "- Initialize variables y_lower, and y_step\n",
    "    - y_lower is the lower bound on the x-axis for the first cluster's silhouette\n",
    "    - y_step is the distance between cluster silhouettes\n",
    "- Initialize variable, breaks\n",
    "    - breaks are the middle point of each cluster silhouette and will be used to\n",
    "      position the axis label\n",
    "- Interate through each cluster label, for each cluster:\n",
    "    - Calcaluate the variable y_upper by adding the number of samples\n",
    "    - Fill the area between y_lower and y_upper using the silhoutte scores for\n",
    "      each sample\n",
    "    - Calculate middle point of y distance. Append the variable break.\n",
    "    - Calculate new y_lower value\n",
    "- Label axes with appropriate names and tick marks\n",
    "- Create dashed line at the average silhouette score over all samples\n",
    "```\n",
    "\n",
    "**Hint**: you might find [ax.fill_betweenx()](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.fill_betweenx.html)\n",
    "and [ax.set_yticks()](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.set_yticks.html?highlight=set_yticks#matplotlib.axes.Axes.set_yticks)/\n",
    "[ax.set_yticklabels()](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.set_yticklabels.html?highlight=set_yticklabels#matplotlib.axes.Axes.set_yticklabels) useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def silhouette_plot(X, y, ax=None):\n",
    "    \"\"\"\n",
    "    Plot silhouette scores for all samples across clusters. \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    X : numpy.ndarray\n",
    "        Numerical data used to cluster the data.\n",
    "    y : numpy.ndarray\n",
    "        Cluster labels assigned to each sample.\n",
    "    ax : matplotlib.Axes\n",
    "        Axis object to plot scores onto. Default is None, and a new axis will\n",
    "        be created.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    matplotlib.Axes\n",
    "    \"\"\"\n",
    "    if ax is None:\n",
    "        ax = plt.subplot()\n",
    "    scores = metrics.silhouette_samples(X, y)\n",
    "    clusters = sorted(np.unique(y))\n",
    "    score_min = 0\n",
    "    y_lower, y_step = 5, 5\n",
    "    props = plt.rcParams['axes.prop_cycle']\n",
    "    colors = itertools.cycle(props.by_key()['color'])\n",
    "    breaks = []\n",
    "    for each, color in zip(clusters, colors):\n",
    "        # Aggregate the silhouette scores for samples, sort scores for\n",
    "        # area filling\n",
    "        cluster_scores = scores[y == each]\n",
    "        cluster_scores.sort()\n",
    "        y_upper = y_lower + len(cluster_scores)\n",
    "        ax.fill_betweenx(np.arange(y_lower, y_upper),\n",
    "                        score_min, cluster_scores,\n",
    "                        facecolor=color, edgecolor=color, alpha=0.7,\n",
    "                        label=each)\n",
    "        breaks.append((y_upper + y_lower) / 2)\n",
    "        # Compute the new y_lower for next plot\n",
    "        y_lower = y_upper + y_step\n",
    "    plt.legend()\n",
    "    ax.set_xlabel(\"Silhouette Coefficient\")\n",
    "    ax.set_ylabel(\"Cluster\")\n",
    "\n",
    "    # Vertical line for threshold \n",
    "    ax.set_yticks(breaks)\n",
    "    ax.set_yticklabels(clusters)\n",
    "    ax.axvline(x=0, linestyle=\"-\", linewidth=2, c='black')\n",
    "    ax.axvline(x=np.mean(scores), linestyle='--', c='black')\n",
    "    plt.tight_layout()\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3C: Put it all together!\n",
    "\n",
    "**Requirements**\n",
    "- Create a function `cluster_and_plot` that will cluster a provided dataset for a range of k-values\n",
    "- The function should return a single figure with two subplots:\n",
    "    - An elbow plot with the \"best\" K value distinguished\n",
    "    - A silhouette plot associated with clustering determined by the provided K value.\n",
    "- Appropriate axes labels\n",
    "\n",
    "**Hint**: You will likely find [plt.subplots()](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.subplots.html?highlight=subplots#matplotlib.pyplot.subplots) useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_and_plot(X, best=3, kmax=10):\n",
    "    \"\"\"\n",
    "    Cluster samples using KMeans and display the results.\n",
    "    \n",
    "    Results are displayed in a (1 x 2) figure, where the\n",
    "    first subplot is an elbow plot and the second subplot\n",
    "    is a silhouette plot.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "        X : (numpy.ndarray)\n",
    "            A (sample x features) data matrix used to cluster\n",
    "            samples.\n",
    "        best : int, optional\n",
    "            Final value of K to use for K-Means clustering.\n",
    "            Default is 3.\n",
    "        kmax : int, optional\n",
    "            Maximum number of clusters to plot in the elbow\n",
    "            plot. Default is 10.\n",
    "    Returns\n",
    "    -------\n",
    "        matplotlib.Figure\n",
    "            Clustering results.\n",
    "    \"\"\"\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2)\n",
    "    scores = np.array([cluster_data(X, k)[1] for k in np.arange(1, kmax+1)])\n",
    "    y, score = cluster_data(X, best)\n",
    "    elbow_plot(np.arange(1, kmax + 1), scores, best, axes[0])\n",
    "    silhouette_plot(X, y, axes[1])\n",
    "    return fig"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:viz]",
   "language": "python",
   "name": "conda-env-viz-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
